### PR TITLE
[TS] LPS-170629 

### PR DESCRIPTION
**Pull request by ricardocousolr**
Motivation
=======
This is a solution to [LPS-170629](https://liferay.atlassian.net/browse/LPS-170629).
    
Some arguments of the StructuredContent API, like `relatedContents`, are readOnly, so one cannot specify related contents when adding or modifying something like a web content. 
    
A review of which arguments of the StructuredContent was done in [LPS-171394](https://liferay.atlassian.net/browse/LPS-171394) with the conclusion to modify:
    -  `relatedContents`
    -  `structuredContentFolderId`
    
Proposed Solution
========
Remove the `readOnly: true` specification for both `relatedContents` and `structuredContentFolderId` within `StructuredContent`.
    
Since `RelatedContent` has its own internal properties (`contentType`, `id`, and `title`), the first two (`contentType` and `id`) need to have the `readOnly: true` attribute removed, too.
    
Finally, when creating the JournalArticle from the StructuredContent information in StructuredContentResourceImpl.java`, the `RelatedContents` and the `structuredContentFolderId` need to be taken into account.
  
(Note: A `RelatedContent` has a `contentType` that needs to be translated into a `className`. For example `StructuredContent` is translated as `JournalArticle`. This map is explicitly built with `_assetTypeTypeToClassNames` similarly to what is done in `TaxonomyVocabularyResourceImpl.java`.)
  
The tests for StructuredContent need to be adapted now that `relatedContents` and `structuredContentFolderId` can be modified. To test the `relatedContents` I added a `blogsEntry`. For the `structuredContentFolderId` I could use the existing `JournalFolder`'s in the involved test classes.

cc @ruben-pulido 
    
Steps to Verify
========
See reproducing steps in [LPS-170629](https://liferay.atlassian.net/browse/LPS-170629).

Checklist (optional)
========

CODE:
- [x] I have considered breaking this PR into smaller PRs if the amount of changed code is too large
- [x] I have performed a self-review of my own code following Liferay's code conventions
- [x] I have checked other areas of the product for solutions to similar problems
- [x] I have added tests that prove my fix is effective or that my feature works

COMMITS:
- [x] My commits organize changes in coherent units
- [ ] There are no commits which should be combined into others
- [x] My commits are ordered in a way that ease understanding
- [x] My commit messages are well formatted and concisely describe what was changed and why it was changed

PR TITLE & DESCRIPTION:
- [x] My PR title includes an issue number and a descriptive title
- [x] My PR description explains the motivation for this change
- [x] My PR description explains the proposed solution
- [x] My PR description includes the steps to verify the change (points to LPS)
- [x] My PR description includes references to other places where this solution is used (if applies)
- [ ] My PR description includes visual aids to understand what has changed (if applies)
- [ ] My PR description explains alternative approaches considered and why they were discarded (if applies)

### CODE REVIEW

**Comment by ruben-pulido** [comment]
ci:test:sf

**Comment by ruben-pulido** [comment]
ci:test:relevant

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/StructuredContentResourceImpl.java [hunk 1]
```diff
@@ -778,6 +795,8 @@ private ServiceContext _createServiceContext(
 
 		serviceContext.setAssetPriority(assetPriority);
 
+		serviceContext.setAssetLinkEntryIds(assetLinkEntryIds);
```
Comment (ruben-pulido) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/StructuredContentResourceImpl.java:806]: Should we sort it alphabetically?

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/StructuredContentResourceImpl.java [hunk 2]
```diff
@@ -821,6 +840,53 @@ private Long[] _getAssetCategoryIds(
...
+
+	private long[] _getAssetLinkEntryIds(RelatedContent[] relatedContents) {
+		List<Long> assetLinkEntryIds = new ArrayList<>();
+
+		if (relatedContents == null) {
+			return ArrayUtil.toLongArray(assetLinkEntryIds);
+		}
+
+		for (RelatedContent relatedContent : relatedContents) {
+			String className = _assetTypeTypeToClassNames.get(
```
Comment (ruben-pulido) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/StructuredContentResourceImpl.java:884]: Can be inlined

**Review by ruben-pulido — CHANGES_REQUESTED** [review]
Hi @ricardocousolr 
I added some minor comments.

baseline also needs to be run on the headless-api module to avoid semantic versioning test failing

**Comment by ricardocousolr** [comment]
Hi @ruben-pulido,

I added the SemVer change and fixed the ordering issue. 

The inlining is not possible though, unless I'm not understanding correctly.

Thanks!

**Comment by ruben-pulido** [comment]
Hi @ricardocousolr 

Thanks for the changes! 

With the inline comment I meant that instead of defining:
```
String className = _assetTypeTypeToClassNames.get(
    relatedContent.getContentType());
```
you can directly compute it when calling `fetchEntry`:

```
AssetEntry assetEntry = _assetEntryLocalService.fetchEntry(
	_assetTypeTypeToClassNames.get(relatedContent.getContentType()),
	 relatedContent.getId());
```

**Comment by ricardocousolr** [comment]
Thanks for the clarification, @ruben-pulido!

I've updated the PR with the change in https://github.com/liferay-echo/liferay-portal/pull/12755/commits/c74e3966d6214ef6d7faf0ca11c2728eb292cdff. 

Let me know if there's anything else to improve or correct.

**Comment by ruben-pulido** [comment]
ci:forward

**Comment by ricardocousolr** [comment]
@ruben-pulido I just fixed the last commit because it had a new line at end of file source formatting error.

**Comment by lfbesada** [comment]
ci:forward

**Comment by ruben-pulido** [comment]
Hi @ricardocousolr!
Can you rerun `gw buildREST' and update the PR?
Thanks!

**Comment by ricardocousolr** [comment]
Hi @ruben-pulido,

I've updated the PR after reruning `gw buildREST`.

**Comment by ruben-pulido** [comment]
ci:forward

**Comment by Tim-Cao** [comment]
See https://github.com/brianchandotcom/liferay-portal/pull/137741

**Comment by Tim-Cao** [comment]
ci:close

**Pull request closed without merging**